### PR TITLE
ENH+BF: S3 and downloaders: support anonymous access + interface to add new providers

### PR DESCRIPTION
**Pull request by yarikoptic**
- [x] Basic and Digest HTTP Auth should work now
- [x]  In interactive session would now allow to establish a new provider configuration for which would get saved into a user directory
- [x] unittests for `process_www_authenticate` etc
- [x] Test at least for the major portions of the UI to enter the new provider/credential
- [ ] may be would include or just be part of integration test above to test Basic and Digest auth on https://jigsaw.w3.org/HTTP/Basic and https://jigsaw.w3.org/HTTP/Digest
- [x] may be more RFing for credentials (allow to make new one persistent etc)
so marking WIP for now

### CODE REVIEW

File: datalad/downloaders/base.py [hunk 1]
```diff
@@ -75,11 +75,12 @@ def __init__(self, credential=None, authenticator=None):
         authenticator: Authenticator, optional
           Authenticator to use for authentication.
         """
+        # Allow for anonymous connection if no credential is provided
```
Comment (kyleam) [inline datalad/downloaders/base.py]: Is this comment misplaced?  The code immediately below is unchanged and doesn't check `credential`.  Perhaps it should be above the `if not credential` line below?

File: datalad/ui/base.py [hunk 2]
```diff
@@ -31,10 +31,21 @@ def question(self, text,
         pass
 
     def yesno(self, *args, **kwargs):
+        # Provide some default sugaring
+        default = kwargs.pop('default', None)
+        if default is not None:
+            if default in {True}:
```
Comment (kyleam) [inline datalad/ui/base.py]: If you're going to have only one value, why not `is True` here and `is False` below?

File: datalad/ui/tests/test_base.py [hunk 3]
```diff
@@ -85,3 +88,6 @@ def nothing(x, k=1):
     @with_testsui(responses='a')
     def ask():
         assert_equal(ui.question('what is a?'), 'a')
+
+
+    ask()
```
Comment (kyleam) [inline datalad/ui/tests/test_base.py]: AFAICS, this fix and the cosmetic changes above are unrelated to yesno default change, which isn't actually covered in the tests.

File: datalad/downloaders/base.py [hunk 4]
```diff
@@ -185,40 +194,67 @@ def access(self, method, url, allow_old_session=True, **kwargs):
                         if not ui.is_interactive:
                             lgr.error(
                                 "Interface is non interactive, so we are "
-                                "reraising: %s" % exc_str(e))
+                                "re-raising: %s" % exc_str(e))
```
Comment (kyleam) [inline datalad/downloaders/base.py]: I think reraising was fine.

File: datalad/downloaders/base.py [hunk 5]
```diff
@@ -228,42 +264,56 @@ def _enter_credentials(self, url, denied_msg, new=True):
...
+                                        " No authenticator is known, cannot "
+                                        "set "
+                                        "any credential")
+                # we could try to just
+                credential_type = self.authenticator.DEFAULT_CREDENTIAL_TYPE
+                if not credential_type:
+                    raise DownloadError(
+                        title +
+                        " %s does not have a default credential type, "
+                        "cannot authenticate"
```
Comment (kyleam) [inline datalad/downloaders/base.py]: Why put "any credential" on a new line?

File: datalad/downloaders/providers.py [hunk 6]
```diff
@@ -337,6 +367,119 @@ def get_provider(self, url, only_nondefault=False):
...
+        if url:
+            ri = RI(url)
+            for f in ('hostname', 'name'):
+                try:
+                    # might need sanitarization
+                    name = str(getattr(ri, f))
+                except AttributeError:
+                    pass
+        known_providers_by_name = {p.name: p for p in self._providers}
+        PROVIDERS_USER_DIR = self._get_providers_dirs()['user']
```
Comment (kyleam) [inline datalad/downloaders/providers.py]: This is local to this function.  Why the allcaps?

File: datalad/downloaders/providers.py [hunk 7]
```diff
@@ -337,6 +367,119 @@ def get_provider(self, url, only_nondefault=False):
...
+                title="New provider name",
+                text="Unique name to identify 'provider' for %s" % url,
+                default=name
+            )
+            filename = pathjoin(PROVIDERS_USER_DIR, '%s.cfg' % name)
+            if name in known_providers_by_name:
+                if ui.yesno(
+                    title="Known provider %s" % name,
+                    text="Provider with name %s already known. Do you want to "
+                         "use it for this session? If 'No' - enter another name"
```
Comment (kyleam) [inline datalad/downloaders/providers.py]: This could be misinterpreted as "To answer 'NO', enter another name here".

File: datalad/downloaders/providers.py [hunk 8]
```diff
@@ -337,6 +367,119 @@ def get_provider(self, url, only_nondefault=False):
...
+            elif path.lexists(filename):
+                ui.error(
+                    "File %s already exists, choose another name" % filename)
+            else:
+                break
+
+        url_re = re.escape(url)
+        while True:
+            url_re = ui.question(
+                title="New provider regular_expression",
```
Comment (kyleam) [inline datalad/downloaders/providers.py]: Drop the underscore?

Comment (yarikoptic) [reply datalad/downloaders/providers.py]: indeed. thanks!

File: datalad/downloaders/providers.py (hunk 7 above)
Comment (yarikoptic) [reply datalad/downloaders/providers.py]: indeed...  I better move this trailing sentence to avoid confusion - I think it will be ok without it

File: datalad/downloaders/providers.py (hunk 6 above)
Comment (yarikoptic) [reply datalad/downloaders/providers.py]: I finally found CAPSLOCK useful! ;-)  I guess leftover from when I hoped to have those just a process-wide constants... will lowercase

File: datalad/downloaders/base.py (hunk 5 above)
Comment (yarikoptic) [reply datalad/downloaders/base.py]: what `"any credential"`? ah -- way above! will fix and compress here a bit. thanks

File: datalad/downloaders/base.py (hunk 4 above)
Comment (yarikoptic) [reply datalad/downloaders/base.py]: ok, will change throughout the codebase... i.e. in 1 more place (a comment ;-))

File: datalad/downloaders/base.py (hunk 1 above)
Comment (yarikoptic) [reply datalad/downloaders/base.py]: I guess it was overall like a TODO comment reminder... I just removed it now

File: datalad/ui/base.py (hunk 2 above)
Comment (kyleam) [reply datalad/ui/base.py]: I find it less readable because I'm not used to seeing it.  As for the logic, I don't consider changing `is True` to `in {True, X}` if the need arises to be a maintenance burden.  And it seems you could use the same logic for pretty much any `is ATOM`, or any `== ATOM` for that matter.  Anyway, my comment was a "just curious" one as well.

File: datalad/ui/tests/test_base.py (hunk 3 above)
Comment (kyleam) [reply datalad/ui/tests/test_base.py]: > This one just spotted

That's fine, but it'd be easy to put that in a separate commit.  When reviewing, I find it confusing when unrelated bits are tossed in (including cosmetic fixups that just add noise to the diffs).

File: datalad/downloaders/http.py [hunk 9]
```diff
@@ -65,6 +65,17 @@
 __docformat__ = 'restructuredtext'
 
 
+def process_www_authenticate(v):
```
Comment (yarikoptic) [inline datalad/downloaders/http.py]: Could you please add a short docstring to describe the purpose of the function and why output should be a list? (that is should provide value for the `supported_types`)

File: datalad/downloaders/tests/test_providers.py [hunk 10]
```diff
@@ -146,3 +151,78 @@ def test_Providers_from_config__files(sysdir, userdir, dsdir):
...
+            providers.enter_new(url, auth_types=["http_basic_auth"])
+        auth_types()
+        ok_exists(op.join(providers_dir, "foo2.cfg"))
+
+        @with_testsui(responses=["foo3", "doesn't match", url_re, auth_type,
+                                 creds, "yes"])
+        def nonmatching_url():
+            providers.enter_new(url, auth_types=["http_basic_auth"])
+        nonmatching_url()
+        ok_exists(op.join(providers_dir, "foo3.cfg"))
```
Comment (yarikoptic) [inline datalad/downloaders/tests/test_providers.py]: Lovely!
I guess the next one would be a similar one for `_enter_credentials`?

File: datalad/downloaders/http.py (hunk 9 above)
Comment (kyleam) [reply datalad/downloaders/http.py]: Uhh you wrote that code.  I don't know its purpose or why its output should be a list, but, sure, I could guess and add a docstring.  In the time it took you to "review" the function you wrote, you could have wrote that docstring.

File: datalad/downloaders/tests/test_providers.py (hunk 10 above)
Comment (kyleam) [reply datalad/downloaders/tests/test_providers.py]: I was hoping I had already fulfilled your request to help with your PR.

File: datalad/downloaders/http.py (hunk 9 above)
Comment (yarikoptic) [reply datalad/downloaders/http.py]: ah, damn me -- should've added a docstring indeed, will do

File: datalad/downloaders/tests/test_providers.py (hunk 10 above)
Comment (yarikoptic) [reply datalad/downloaders/tests/test_providers.py]: you did help! ok,I could continue whenever I get a chance, but not sure when it would be :-(

Comment (kyleam) [reply datalad/downloaders/tests/test_providers.py]: I wasn't saying I wouldn't help more, just stating what I had hoped :)

I'll look into filling out the tests more, starting with `_enter_credentials`.

Comment (yarikoptic) [reply datalad/downloaders/tests/test_providers.py]: thanks in advance! ;-)

**Comment by yarikoptic** [comment]
@kyleam pushed two commits to "enable" access to the s3 buckets with period in their name (apparently needs special handling and now there is openneuro.org bucket)

**Comment by yarikoptic** [comment]
Thank you very much @kyleam

**Comment by yarikoptic** [comment]
Looks great. Let's see how it holds to real life testing
